「Discord API」の利用マニュアルを追加 (#81)
gimKondo/sound-of-cthulhu: MANUAL.md

# Sound of Cthulhu マニュアル

## Discord APIの利用方法

### DiscordでBotアプリケーションを作成

- Botアプリケーションを作成(Discordのリファレンスを参照)
  - <https://discordjs.guide/preparations/setting-up-a-bot-application.html#creating-your-bot>
  - <https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links>
- Botには下記権限を与えること
  - GENERAL PERMISSIONS
    - View Channel
  - Text Permissions
    - Send Messages
  - Voice Permissions
    - Connect
    - Speak
- 作成したアプリケーションがサーバ上に現れ、Tokenが取得出来る事を確認する。  
  - 招待リンクはこのようになる。  

>   - `https://discord.com/api/oauth2/authorize?client_id={YOUR_CLIENT_ID}&permissions=3148800&scope=bot`

Reviewer: 単に私が分かってないだけなのですが、ユーザはこのリンクを叩いて(DiscordのUI上で入力？)、Tokenを取得するということでしょうか？

Author: この方法で、アプリケーションを新規作成します。
https://discordjs.guide/preparations/setting-up-a-bot-application.html#your-token
そうすると、そのアプリケーションに紐付いたTOKENが取得できます。
この時、botが必要とする権限によって、リンクが変わってくるので、そのリンクを叩いてワークスペースにアプリケーションをインストールする流れとなります。
（discord.pyのドキュメントですが、こっちの方がわかりやすいんですよね・・・）
https://discordpy.readthedocs.io/ja/latest/discord.html

Reviewer: 分かりやすい説明ありがとうございます！
ただ、そうするとユーザがやるべきことは
「Tokenが取得出来る事を確認する」
ではなくて、
「Tokenを取得する」
「Botユーザをサーバに招待する」
の2つですかね？

Author: あ、そうなりますね。
上に記載しているリファレンスでも、分けて記載しておかないと伝わらなかったですね。
「アプリケーションを作成する」と「サーバにインストールする」は別項目でした。

Author: 3c5ca20 に追加しました

> ## Discord APIの利用方法

Reviewer: ```suggestion
# Sound of Cthulhu マニュアル

## Discord APIの利用方法
```

Author: あ、このMANUALに追記していく感じですね
了解です

Reviewer: - ルートにあるMANUAL.mdなので、Discord関連の内容ではなく、全体のマニュアルなんですよね？
  - これまでは説明が必須なほど複雑な操作が無かったのと、面倒だったのとで作ってなかったですが
- h2から始めていたのは、さらに上の見出しが存在しているが、書き忘れなのではないかと判断しました

> ### 権限
> 
> アプリケーション作成時に権限が必要になるので、控えておく。
> 
> #### 権限
> - GENERAL PERMISSIONS
>   - View Channel
> - Text Permissions
>   - Send Messages
> - Voice Permissions
>   - Connect
>   - Speak
> 
> ### DiscordでBotアプリケーションを作成
> 
> 以下のURLを参照し、アプリケーションを作成。  
> https://discordjs.guide/preparations/setting-up-a-bot-application.html#creating-your-bot  
> https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links  
>   
> 作成したアプリケーションがサーバ上に現れ、Tokenが取得出来る事を確認する。  
> 招待リンクはこのようになる。  
> https://discord.com/api/oauth2/authorize?client_id={YOUR_CLIENT_ID}&permissions=3148800&scope=bot

Reviewer: ```suggestion
### DiscordでBotアプリケーションを作成

- Botアプリケーションを作成(Discordのリファレンスを参照)
  - <https://discordjs.guide/preparations/setting-up-a-bot-application.html#creating-your-bot>
  - <https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links>
- Botには下記権限を与えること
  - GENERAL PERMISSIONS
    - View Channel
  - Text Permissions
    - Send Messages
  - Voice Permissions
    - Connect
    - Speak
- 作成したアプリケーションがサーバ上に現れ、Tokenが取得出来る事を確認する。  
  - 招待リンクはこのようになる。  
  - `https://discord.com/api/oauth2/authorize?client_id={YOUR_CLIENT_ID}&permissions=3148800&scope=bot`
```

Author: 最初この順番で書いてたんですが、権限が最初にわかってないとアプリケーションの作成が難しいな？と思ってしまったんですよね
わかりそうであればいいんですが、直しましょうか？

Reviewer: - 「権限」という見出しの下に「権限」という小見出しがあって、不要な構造に思えたので統合
- 何の説明もなく「権限」と書かれても何の話なのか分からないのでアプリケーションの作成の下に移動
- 「控えておく」とあるが、ここに書かれているので読者がどうアクションを取れば良いか不明な手順だったので削除
  - アプリケーションを作成する際にここを参照して設定するのですよね？

あたりを意図しています。

Author: あ、権限の説明は必要でした。
そうして、改めてみたらこの記載だと良さそうですね。
ありがとうございます、こちらにしておきます。

### Tokenの設定方法

1. 上記のアプリケーション作成時に取得したTokenを控えておく
2. デバイスリストから「Discord API」を選択
3. ダイアログが出て、表示されたパスに設定用のiniファイルが作成される
4. iniファイルにTokenを入力するようになっているので、テキストエディタでTokenを入力する。
5. 再度、デバイスリストから「Discord API」を選択

### アプリケーションをチャンネルに追加する方法

1. Discordのユーザでボイスチャンネルに参加する。

> 2. テキストチャンネルで「:soc: join」と入力する。

Reviewer: `:soc:` という名前はBotを作るときに決めるのではないかと思うのですが、どうでしょう？

Author: https://github.com/gimKondo/sound-of-cthulhu/blob/feat/add-discord-document/src/background.js#L107-L116
あー、本当はそうしたかったんですが、messageから取得できる条件で決める必要があり
この段階では文字列決め打ちにせざるを得なかった感じですね。
（基本的なチュートリアルなどでもそうなっていたので）
なにか妙案があればいいのですが

Reviewer: 伝わっていないようなので補足します。
`:soc:` で反応するようにbotアプリケーションを作る際にbot名(？)か何かを決める必要があるのではないかと解釈しました。
botの設定に関わらず、 `:soc: join` とDiscordに打ち込むと、必要なbotがjoinされるということでしょうか？

Author: >botの設定に関わらず、 :soc: join とDiscordに打ち込むと、必要なbotがjoinされるということでしょうか？

はい、そうですね
トリガー文字列はどういうものでも問題ありません。
（当初はアプリケーション名をCoCBotという名前で動かしていました）
botの設定での動作は権限のみが影響します。

Reviewer: あー、理解しました。
Discordサーバに登録しているbotをこのメッセージで招待しているのではなく、

1. SoCがメッセージを監視している
2. メッセージを打ち込む
3. `:soc: join` だった場合にトークンに対応したbotをチャンネルに参加させる

という動きになっているんですね。

Author: あ、そうなります。
そうか、そのあたり別途記載必要でしょうかねー。

Author: botを招待出来れば良いのですが、TOKENが必要になってしまうのでいい方法が思いついてない現状ですね

Reviewer: > そうか、そのあたり別途記載必要でしょうかねー。

いえ、単純に私が理解していなくて、これで正しいのか判断付かなかっただけです。
ユーザ向けの「使い方」の記載としては不要かと思います。

> botを招待出来れば良いのですが、TOKENが必要になってしまうのでいい方法が思いついてない現状ですね

特にマズいと思ってないのですが、現状の方法だと何か懸念がありますか？

Author: とても細かいのですが、KPとしてsound-of-cthulhuを使うユーザの分だけアプリケーションが増えてしまうぐらいでしょうか。
アプリケーションを追加出来るのが管理者権限持ちだけなので、一般ユーザがsound-of-cthulhuを使えなかったりします。
とはいえ、このあたりは仕様上仕方ないかなと思っていますので、ドキュメント外の問題ですね。

Reviewer: いや、そこはユーザが影響受ける事項なので書きましょう。
開発側としては「仕様上仕方ない」で良いのですが、それはユーザには関係ないことですし。

「Discord APIの使い方」の最初に、 `この機能を使うにはDiscordサーバの管理者権限が必要です。`  と書いておくのはどうでしょう？

Author: なるほど、確かにそうですね。
ご提案通り、最初の所に記載するようにします。

Author: 6e94e9a で追加しました

#### 注意事項

> 現在、custom_emojiで「:soc:」を登録してjoinコマンドを発行しても入力を受け付けない。  

Reviewer: 素朴な疑問ですが、craigはこれでも動いてるのは何故ですかね？

Author: あ、昨日は気が付かなかったんですが正規表現で取ってるのかな
https://github.com/Yahweasel/craig/blob/463ce8b892205ff1e26a6554120f7d931d79028a/craig/rec.js#L1014-L1015
まぁ、ドキュメント以外の所で修正ですかねー

Reviewer: まー、大きな問題ではないので、そのうち修正くらいで良いかと！

# Sound of Cthulhu マニュアル

## Discord APIの利用方法

この機能を使うにはDiscordワークスペースの管理者権限が必要です。

### DiscordでBotアプリケーションの設定

- Botアプリケーションを作成(Discordのリファレンスを参照)
  - <https://discordjs.guide/preparations/setting-up-a-bot-application.html#creating-your-bot>
  - <https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links>

> - 作成したアプリケーションがサーバ上に現れ、Tokenが取得出来る事を確認する。

Reviewer: 「確認する」ではなく、取得して控えておく方が良さそうですね。
```suggestion
- 作成したアプリケーションがサーバに現れるので、Tokenを取得して控えておく。
```

  - [SETTING]
    - [Bot]
      - [Copy] クリップボードにTokenがコピーされる
- 招待リンクを作成し、サーバ上にアプリケーションを追加する。
  - [SETTING]
    - [OAuth2]
      - [SCOPES]
        - [bot] にチェックを入れ以下の権限を与える
          - GENERAL PERMISSIONS
            - View Channel
          - Text Permissions
            - Send Messages
          - Voice Permissions
            - Connect
            - Speak
  - 上記画面に招待リンクが作成される
    - 招待リンクはこのようになる。
      - `https://discord.com/api/oauth2/authorize?client_id={YOUR_CLIENT_ID}&permissions=3148800&scope=bot`

### Tokenの設定方法

> 1. 上記のアプリケーション作成時に取得したTokenを控えておく

Reviewer: ```suggestion
1. アプリケーション作成時に取得しておいたTokenを用意する。
```
アプリケーション作成の手順修正に合わせて、こちらの文言も修正。
